### ISSUE TITLE
new install in new project - multiple Typescript errors (TS1005, TS1128, TS1109)

### ISSUE THREAD

**Issue by johnpankowicz** [nartc/mapper#256]
I  created a new Angular project with "ng new" and then ran:
```
npm i @automapper/core
npm i @automapper/classes
npm i reflect-metadata
npm i -D @automapper/types
```
I added the following line to the top of app.component.ts:
`import { AutoMap } from '@automapper/classes';`

When I run "npm start", I immediately get the following errors:

```
ERROR in node_modules/@automapper/types/src/plugin.d.ts:32:210 - error TS1005: ',' expected.

32     preMap?<TSource extends Dictionary<TSource> = any, TDestination extends Dictionary<TDestination> = any>(source: TKey, destination: TKey, sourceObj?: TSource, destinationObj?: TDestination): [sourceInstance: TSource, destinationInstance: TDestination];

                                                                                      ~
node_modules/@automapper/types/src/plugin.d.ts:32:240 - error TS1005: ',' expected.

32     preMap?<TSource extends Dictionary<TSource> = any, TDestination extends Dictionary<TDestination> = any>(source: TKey, destination: TKey, sourceObj?: TSource, destinationObj?: TDestination): [sourceInstance: TSource, destinationInstance: TDestination];

                                                                                                                    ~
node_modules/@automapper/types/src/core.d.ts:24:26 - error TS1005: ',' expected.

24     preConditionPredicate: (source: TSource) => boolean,
                            ~
node_modules/@automapper/types/src/core.d.ts:25:18 - error TS1005: ',' expected.

25     defaultValue?: TSelectorReturn
                    ~
node_modules/@automapper/types/src/core.d.ts:31:9 - error TS1005: ',' expected.

31     type: TransformationType.MapDefer,
           ~
node_modules/@automapper/types/src/core.d.ts:32:7 - error TS1005: ',' expected.

32     fn: DeferFunction<TSource, TDestination, TSelectorReturn>
         ~
node_modules/@automapper/types/src/core.d.ts:35:9 - error TS1005: ',' expected.

35     type: TransformationType.MapFrom,
           ~
node_modules/@automapper/types/src/core.d.ts:36:7 - error TS1005: ',' expected.

36     fn: (source: TSource, destination?: TDestination) => TSelectorReturn,
         ~
node_modules/@automapper/types/src/core.d.ts:37:9 - error TS1005: ',' expected.

37     misc: ValueSelector<TSource, TDestination, TSelectorReturn> | null
           ~
node_modules/@automapper/types/src/core.d.ts:40:9 - error TS1005: ',' expected.

40     type: TransformationType.MapWith,
           ~
node_modules/@automapper/types/src/core.d.ts:41:7 - error TS1005: ',' expected.

41     fn: (sourceObj: TSource, mapper: Mapper) => TSelectorReturn | undefined | null,
         ~
node_modules/@automapper/types/src/core.d.ts:42:9 - error TS1005: ',' expected.

42     misc: ValueSelector<TSource>
           ~
node_modules/@automapper/types/src/core.d.ts:45:9 - error TS1005: ',' expected.

45     type: TransformationType.Condition,
           ~
node_modules/@automapper/types/src/core.d.ts:46:7 - error TS1005: ',' expected.

46     fn: (source: TSource, ...sourceMemberPaths: string[]) => TSelectorReturn
         ~
node_modules/@automapper/types/src/core.d.ts:48:188 - error TS1005: ',' expected.

48 export declare type FromValueReturn<TSource extends Dictionary<TSource> = any, TDestination extends Dictionary<TDestination> = any, TSelectorReturn = SelectorReturn<TDestination>> = [type: TransformationType.FromValue, fn: () => TSelectorReturn];

                                                                ~
node_modules/@automapper/types/src/core.d.ts:48:222 - error TS1005: ',' expected.

48 export declare type FromValueReturn<TSource extends Dictionary<TSource> = any, TDestination extends Dictionary<TDestination> = any, TSelectorReturn = SelectorReturn<TDestination>> = [type: TransformationType.FromValue, fn: () => TSelectorReturn];

                                                                                                  ~
node_modules/@automapper/types/src/core.d.ts:50:9 - error TS1005: ',' expected.

50     type: TransformationType.ConvertUsing,
           ~
node_modules/@automapper/types/src/core.d.ts:51:7 - error TS1005: ',' expected.

51     fn: Selector<TSource, TConvertSource>
         ~
node_modules/@automapper/types/src/core.d.ts:54:9 - error TS1005: ',' expected.

54     type: TransformationType.NullSubstitution,
           ~
node_modules/@automapper/types/src/core.d.ts:55:7 - error TS1005: ',' expected.

55     fn: (source: TSource, ...sourceMemberPaths: string[]) => TSelectorReturn
         ~
node_modules/@automapper/types/src/core.d.ts:57:41 - error TS1005: ',' expected.

57 export declare type IgnoreReturn = [type: TransformationType.Ignore];
                                           ~
node_modules/@automapper/types/src/core.d.ts:59:9 - error TS1005: ',' expected.

59     type: TransformationType.MapWithArguments,
           ~
node_modules/@automapper/types/src/core.d.ts:60:7 - error TS1005: ',' expected.

60     fn: (source: TSource, extraArguments: Record<string, unknown>) => TSelectorReturn
         ~
node_modules/@automapper/types/src/core.d.ts:63:9 - error TS1005: ',' expected.

63     type: TransformationType.MapInitialize,
           ~
node_modules/@automapper/types/src/core.d.ts:64:7 - error TS1005: ',' expected.

64     fn: Selector<TSource, TSelectorReturn>
         ~
node_modules/@automapper/types/src/core.d.ts:113:10 - error TS1005: ',' expected.

113     mapFn: MemberMapReturn<TSource, TDestination, TSelectorReturn>,
             ~
node_modules/@automapper/types/src/core.d.ts:114:13 - error TS1005: ',' expected.

114     preCond?: PreConditionReturn<TSource, TDestination, TSelectorReturn>
                ~
node_modules/@automapper/types/src/core.d.ts:117:10 - error TS1005: ',' expected.

117     paths: [target: string, origin?: string],
             ~
node_modules/@automapper/types/src/core.d.ts:117:19 - error TS1005: ',' expected.

117     paths: [target: string, origin?: string],
                      ~
node_modules/@automapper/types/src/core.d.ts:117:36 - error TS1005: ',' expected.

117     paths: [target: string, origin?: string],
                                       ~
node_modules/@automapper/types/src/core.d.ts:118:19 - error TS1005: ',' expected.

118     transformation: MappingTransformation<TSource, TDestination, TSelectorReturn>
                      ~
node_modules/@automapper/types/src/core.d.ts:121:13 - error TS1005: ',' expected.

121     mappings: [source: TSource, destination: TDestination],
                ~
node_modules/@automapper/types/src/core.d.ts:121:22 - error TS1005: ',' expected.

121     mappings: [source: TSource, destination: TDestination],
                         ~
node_modules/@automapper/types/src/core.d.ts:121:44 - error TS1005: ',' expected.

121     mappings: [source: TSource, destination: TDestination],
                                               ~
node_modules/@automapper/types/src/core.d.ts:122:15 - error TS1005: ',' expected.

122     properties: Array<[
                  ~
node_modules/@automapper/types/src/core.d.ts:123:13 - error TS1005: ',' expected.

123         path: string,
                ~
node_modules/@automapper/types/src/core.d.ts:124:24 - error TS1005: ',' expected.

124         mappingProperty: MappingProperty<TSource, TDestination, SelectorReturn<TDestination>>,
                           ~
node_modules/@automapper/types/src/core.d.ts:125:27 - error TS1005: ',' expected.

125         nestedMappingPair?: [unknown, unknown]
                              ~
node_modules/@automapper/types/src/core.d.ts:126:6 - error TS1005: ';' expected.

126     ]>,
         ~
node_modules/@automapper/types/src/core.d.ts:126:7 - error TS1109: Expression expected.

126     ]>,
          ~
node_modules/@automapper/types/src/core.d.ts:127:12 - error TS1005: ';' expected.

127     actions: [
               ~
node_modules/@automapper/types/src/core.d.ts:128:19 - error TS1109: Expression expected.

128         beforeMap?: MapAction<TSource, TDestination>,
                      ~
node_modules/@automapper/types/src/core.d.ts:128:53 - error TS1109: Expression expected.

128         beforeMap?: MapAction<TSource, TDestination>,
                                                        ~
node_modules/@automapper/types/src/core.d.ts:129:18 - error TS1109: Expression expected.

129         afterMap?: MapAction<TSource, TDestination>
                     ~
node_modules/@automapper/types/src/core.d.ts:130:5 - error TS1005: '(' expected.

130     ],
        ~
node_modules/@automapper/types/src/core.d.ts:131:23 - error TS1109: Expression expected.

131     namingConventions?: [source: NamingConvention, destination: NamingConvention],
                          ~
node_modules/@automapper/types/src/core.d.ts:131:32 - error TS1005: ',' expected.
131     namingConventions?: [source: NamingConvention, destination: NamingConvention],
                                   ~
node_modules/@automapper/types/src/core.d.ts:131:63 - error TS1005: ',' expected.

131     namingConventions?: [source: NamingConvention, destination: NamingConvention],
                                                                  ~
node_modules/@automapper/types/src/core.d.ts:132:11 - error TS1109: Expression expected.

132     bases?: [baseSource: unknown, baseDestination: unknown][]
              ~
node_modules/@automapper/types/src/core.d.ts:132:24 - error TS1005: ',' expected.

132     bases?: [baseSource: unknown, baseDestination: unknown][]
                           ~
node_modules/@automapper/types/src/core.d.ts:132:50 - error TS1005: ',' expected.

132     bases?: [baseSource: unknown, baseDestination: unknown][]
                                                     ~
node_modules/@automapper/types/src/core.d.ts:132:61 - error TS1011: An element access expression should take an argument.

132     bases?: [baseSource: unknown, baseDestination: unknown][]

node_modules/@automapper/types/src/core.d.ts:133:1 - error TS1128: Declaration or statement expected.

133 ];
    ~
node_modules/@automapper/types/src/core.d.ts:149:13 - error TS1005: ',' expected.

149     property: string,
                ~
node_modules/@automapper/types/src/core.d.ts:150:15 - error TS1005: ',' expected.

150     metaTypeFn: () => String | Number | Boolean | Date | TMetaType
                  ~
```

**Comment by micalevisk** [comment]
I got these too. I think you should use [`skipLibCheck: true`](https://www.typescriptlang.org/docs/handbook/compiler-options.html) in your `tsconfig.json` file.

**Comment by nartc** [comment]
@johnpankowicz what angular version are you using? I just initialized a new angular app (11.2) and everything runs fine

```shell
ng new test-mapper --minimal --defaults --strict 
npm i @automapper/{core,classes} reflect-metadata
npm i -D @automapper/types
```
The other steps are:
- Add `emitDecoratorMetadata` to `tsconfig`
- Add `import 'reflect-metadata' to `polyfills.ts`

And everything works properly. 
Here's a Angular project (11.2): https://github.com/nartc/ng-mapper-bench

**Comment by johnpankowicz** [comment]
I just noticed: [https://github.com/nartc/mapper/issues/246](https://github.com/nartc/mapper/issues/246)
which appears to indicate that I need Angular 10. I will try upgrading.

**Comment by nartc** [comment]
@johnpankowicz your angular cli is probably of an old version. If your project is simple enough, ng update would definitely work just fine

**Comment by johnpankowicz** [comment]
@nartc Thank for the quick response! Doing the "ng update" fixed it. No more errors now on "npm start".

But what are these steps for?
* Add emitDecoratorMetadata to tsconfig
* Add import 'reflect-metadata' to polyfills.ts

They are not part of the quick start instructions. Where should I have learned about this in the docs? I do not have these changes in tsconfig.json & polyfill.ts.

**Comment by nartc** [comment]
@johnpankowicz Those steps arent in the docs because it wasn’t written with Angular in mind (my apologies, the docs should have had those steps listed). And because Reflect Metadata does call out those steps so I overlooked it. Do you want to submit a PR to the docs for that?

**Comment by johnpankowicz** [comment]
@nartc OK, I can do that. But I rather wait till I have more that just the "import {AutoMap}..." statement working. I haven't tried actually using the code yet. I may find other problems with Angular.

**Comment by johnpankowicz** [comment]
I got the sample code from your [Stackblitz example](https://stackblitz.com/edit/typescript-automapper?file=index.ts) working in my Angular app. I put it into a class that I imported into an Angular component. I did not need to import reflect-metadata. I assume that this is because Angular is already using reflection through core-js.
"emitDecoratorMetadata" was already in tsconfig.json. I didn't notice it before.
Therefore, there is probably no need to update the docs.
BTW: The function "getUserDto" in your stackblitz example is better named "getUser" since it is getting a User object and not a UserDto object.
